aptos-move/move-examples/fungible_asset/coin_wrapper/sources/coin_wrapper.move

```diff
@@ -0,0 +1,220 @@
+        let fa_address_str = string_utils::to_string(&fa_address);
+        // We want to strip the prefix "@"
+        string::sub_string(&fa_address_str, 1, string::length(&fa_address_str))
+    }
+
+    /// Create the coin wrapper account to host all the deposited coins.
+    fun init_module(coin_wrapper_signer: &signer) {
+        let (coin_wrapper_signer, signer_cap) =
+            account::create_resource_account(coin_wrapper_signer, COIN_WRAPPER_NAME);
+        move_to(&coin_wrapper_signer, WrapperAccount {
+            signer_cap,
+            coin_to_fungible_asset: smart_table::new(),
+            fungible_asset_to_coin: smart_table::new(),
+        });
+    }
+
+    /// Wrap the given coins into fungible asset. This will also create the fungible asset wrapper if it doesn't exist
+    /// yet. The coins will be deposited into the main resource account.
+    public(friend) fun wrap<CoinType>(coins: Coin<CoinType>): FungibleAsset acquires WrapperAccount {
+        // Ensure the corresponding fungible asset has already been created.
+        create_fungible_asset<CoinType>();
+
+        // Deposit coins into the main resource account and mint&return the wrapper fungible assets.
+        let amount = coin::value(&coins);
+        aptos_account::deposit_coins(wrapper_address(), coins);
+        let mint_ref = &fungible_asset_data<CoinType>().mint_ref;
+        fungible_asset::mint(mint_ref, amount)
+    }
+
+    /// Unwrap the given fungible asset into coins. This will burn the fungible asset and withdraw&return the coins from
+    /// the main resource account.
+    /// This errors out if the given fungible asset is not a wrapper fungible asset.
+    public(friend) fun unwrap<CoinType>(fa: FungibleAsset): Coin<CoinType> acquires WrapperAccount {
+        let amount = fungible_asset::amount(&fa);
+        let burn_ref = &fungible_asset_data<CoinType>().burn_ref;
+        fungible_asset::burn(burn_ref, fa);
+        let wrapper_signer = &account::create_signer_with_capability(&wrapper_account().signer_cap);
+        coin::withdraw(wrapper_signer, amount)
+    }
+
+    /// Create the fungible asset wrapper for the given CoinType if it doesn't exist yet.
+    public(friend) fun create_fungible_asset<CoinType>(): Object<Metadata> acquires WrapperAccount {
+        let coin_type = type_info::type_name<CoinType>();
+        let wrapper_account = mut_wrapper_account();
+        let coin_to_fungible_asset = &mut wrapper_account.coin_to_fungible_asset;
+        let wrapper_signer = &account::create_signer_with_capability(&wrapper_account.signer_cap);
+        if (!smart_table::contains(coin_to_fungible_asset, coin_type)) {
+            let metadata_constructor_ref = &object::create_named_object(wrapper_signer, *string::bytes(&coin_type));
+            primary_fungible_store::create_primary_store_enabled_fungible_asset(
+                metadata_constructor_ref,
+                option::none(),
+                coin_type,
```

Comment:
the name can go very long, exceeding fungible asset name length limit

Reply:
Coins have the same name length constraint as fungible assets: https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/coin.move#L68

Reply:
coin_type could be 0xaaaaaaaaaaaaaaaaaaaaaaaa...::my_coin::MyCoin that goes beyond 32 since it is `type_info::type_name<T>` rather than `coin::name<CoinType>()`

Reply:
Oh that's true. I should use the coin's name here. Let me update